botorch/acquisition/hentropy_search.py

```diff
@@ -0,0 +1,282 @@
+        self.loss_function_hyperparameters = loss_function_hyperparameters
+        self.loss_function = loss_function_class(
+            **self.loss_function_hyperparameters,
+        )
+
+    @t_batch_mode_transform()
+    def forward(self, X: Tensor, A: Tensor) -> Tensor:
+        r"""Evaluate qHEntropySearch objective (q-HES) on the candidate set `X`.
+
+        Args:
+            X: Design tensor of shape `(batch) x q x num_dim_design`.
+            A: Action tensor of shape `(batch) x n_fantasy_at_design_pts
+                x num_actions x num_dim_action`.
+
+        Returns:
+            A Tensor of shape `(batch)`.
+        """
+
+        # construct the fantasy model of shape `n_fantasy_at_design_pts x b`
+        fantasy_model = self.model.fantasize(
+            X=X, sampler=self.design_sampler, observation_noise=True
+        )
+
+        # Permute shape of A to work with self.model.posterior correctly
+        A = A.permute(1, 0, 2, 3)
+
+        fantasized_outcome = self.action_sampler(fantasy_model.posterior(A))
+        # >>> n_fantasy_at_action_pts x n_fantasy_at_design_pts
+        # ... x batch_size x num_actions x 1
+
+        fantasized_outcome = fantasized_outcome.squeeze(dim=-1)
+        # >>> n_fantasy_at_action_pts x n_fantasy_at_design_pts
+        # ... x batch_size x num_actions
+
+        values = self.loss_function(A=A, Y=fantasized_outcome)
+        # >>> n_fantasy_at_design_pts x batch_size
+
+        # return average over the fantasy samples
+        return values.mean(dim=0)
+
+    def get_augmented_q_batch_size(self, q: int) -> int:
+        r"""Get augmented q batch size for optimization.
+
+        Args:
+            q: The number of candidates to consider jointly.
+
+        Returns:
+            The augmented size for optimization.
+        """
+
+        return q + self.cfg.n_dim_action
+
+    def extract_candidates(self, batch_as_full: Tensor) -> Tensor:
+        r"""We only return X as the set of candidates post-optimization.
+
+        Args:
+            batch_as_full: A `b x (q + num_fantasies) x d`-dim Tensor with `b`
+                t-batches of `q + num_fantasies` design points each.
+
+        Returns:
+            A `b x q x d`-dim Tensor with `b` t-batches of `q` design points each.
+        """
+
+        assert len(batch_as_full.shape) == 2
+        n_restarts = batch_as_full.size(0)
+        split_sizes = [
+            batch_as_full.size(1) - self.cfg.n_actions * self.cfg.n_dim_action,
+            self.cfg.n_actions * self.cfg.n_dim_action,
+        ]
+        batch_xs, batch_as_full = torch.split(batch_as_full, split_sizes, dim=1)
+        batch_xs = batch_xs.reshape(n_restarts, -1, self.cfg.n_dim_design)
+        batch_as_full = batch_as_full.reshape(
+            n_restarts, self.cfg.n_actions, self.cfg.n_dim_action
+        )
+
+        return batch_xs
+
+
+class qLossFunctionTopK(nn.Module):
+    r"""Batch loss function for the task of finding top-K."""
```

Comment:
Might be worth specifying that top-K is relative to the values of the objective function?